### ISSUE TITLE
repoll immediately when any request is unfulfilled

### ISSUE THREAD

**Issue by tnielens** [zio/zio-kafka#664]
In `RunLoop.run`
```scala
  def run: ZIO[Scope, Nothing, Fiber.Runtime[Throwable, Unit]] =
    ZStream
      .mergeAll(3, 1)(
        ZStream(Command.Poll).repeat(Schedule.spaced(pollFrequency)),
        ZStream.fromQueue(requestQueue).mapChunks(c => Chunk.single(Command.Requests(c))),
...
```
the `RunLoop` takes a chunk of partition Requests from the `RunLoop.requestQueue` and runs an iteration of the loop in `handleOperational`.  The Requests that aren't fulfilled remain in the `RunLoop.State.pendingRequests`. But a loop iteration will only run if other Requests are present  on the `RunLoop.requestQueue` or upon the next `Command.Poll` occurrence. That can potentially block the processing for a certain time. For example, if the topic has a single partition, the processing can block for up to the full `pollFrequency` when the input records are spaced by more than the `pollTimeout`.

As a suggestion, maybe the unfulfilled requests could be requeued on `RunLoop.requestQueue`. That would immediately trigger an iteration of the loop.

**Comment by tnielens** [comment]
Relates to #656 and #428.

**Comment by guizmaii** [comment]
> The Requests that aren't unfulfilled

To be sure, do you mean "The Requests that aren't fulfilled"? 🤔

**Comment by tnielens** [comment]
Yes! I've edited the description.

**Comment by guizmaii** [comment]
I wonder if this issue isn't related to this: https://github.com/zio/zio-kafka/issues/428#issuecomment-1445352019 🤔

**Comment by svroonland** [comment]
@erikvanoosten  To continue from https://github.com/zio/zio-kafka/issues/656#issuecomment-1455054058:

As car as I can see there are three reasons why we should execute poll()s:
1. To fetch new data for partitions with pending requests
2. At startup, to establish communication with the broker and get a partition assignment and start requests. We should keep polling during (eager) rebalancing as well, because we first revoke partition streams and there are no requests anymore after that. Also manual topic-partition assignment is a special case. 
3. Periodically (at least < `max.poll.interval.ms`) to signal that the consumer is still alive and should not be removed from the consumer group.

For 3, you could argue that if there are no pending requests for a while, the consuming application is not alive. I think that is the behavior we should aim for.

**Comment by erikvanoosten** [comment]
> For 3, you could argue that if there are no pending requests for a while, the consuming application is not alive.

Yes, that is very true.

So in conclusion:
* Poll once after subscribing, after a re-balance and after manually assigning partitions.
* Poll when there is at least one pending data request.

In addition, we only poll when we are subscribed.

For case 3, we will learn that we lost partitions only during the next call to poll. By then, any pending commits are broken. What do we do at that moment? It would be nice if the application could recover. Should we interrupt the current streams, subscribe again, and then start new streams as partitions are assigned?

**Comment by erikvanoosten** [comment]
We recently had a bug that caused consumers to be not shutdown. These dangling consumers were still processing some partitions but not all. You would expect that the non-used partitions would be re-assigned to other consumers, but that didn't happen because they were paused.

So the whole 'signal that the consumer is still alive'-thing only works when the consumer stopped consuming _all_ partitions. It does not work when just one or a few partitions are no longer consumed from.

**Comment by erikvanoosten** [comment]
> So in conclusion:
>     * Poll once after subscribing, after a re-balance and after manually assigning partitions.
>     * Poll when there is at least one pending data request.

@svroonland I see from the code in #683 that it should be:

 * After subscribing, after a re-balance and after manually assigning partitions, start polling regularly until the first pending data request.
 * Poll when there is at least one pending data request.

WDYT?

**Comment by svroonland** [comment]
That sounds good.

**Comment by svroonland** [comment]
> So the whole 'signal that the consumer is still alive'-thing only works when the consumer stopped consuming all partitions. It does not work when just one or a few partitions are no longer consumed from.

We could keep track of the last requests made for each assigned partition and exit the consumer when we haven't seen requests recently enough, configurable with some timeout setting. Would be good to create a new issue for this.

**Comment by tnielens** [comment]
> We could keep track of the last requests made for each assigned partition and exit the consumer when we haven't seen requests recently enough, configurable with some timeout setting. Would be good to create a new issue for this.

This would reimplements max.poll.timout.ms at the partition level? If going that route I recommend to stay close to the java consumer settings naming, for example max.partition.poll.interval.ms, and build on the users knowledge of the java clients.

**Comment by erikvanoosten** [comment]
@tnielens in #686 I propose to re-use `max.poll.interval.ms` for the setting. The rationale is that polling should happen within that time for each and every partition. If you see a use case where a single stream is allowed to be slower than that, we can introduce another setting but use `max.poll.interval.ms` as default. (Remember, `max.poll.interval.ms` defaults to 5 minutes.)

**Comment by svroonland** [comment]
It looks like there's a 4th need for polling: to get the confirmation for commits (communicated via the `OffsetCommitCallback`). It seems to take several poll() calls sometimes before the callback is called..

**Comment by svroonland** [comment]
Implemented in #683. Thanks @tnielens, @erikvanoosten and @guizmaii for the ideas, discussion and feedback!
